runtime/src/main/cpp/Memory.cpp

```diff
@@ -2056,51 +2056,81 @@ OBJ_GETTER(initSharedInstance,
+  auto* state = memoryState;
+  auto timestamp = state->lastGcTimestamp;
+  return (static_cast<int32_t>(reinterpret_cast<intptr_t>(state))) ^
+    (static_cast<int32_t>(reinterpret_cast<intptr_t>(value))) ^
+    static_cast<int32_t>(timestamp >> 32) ^
+    static_cast<int32_t>(timestamp);
+}
+
 OBJ_GETTER(swapHeapRefLocked,
-    ObjHeader** location, ObjHeader* expectedValue, ObjHeader* newValue, int32_t* spinlock) {
+    ObjHeader** location, ObjHeader* expectedValue, ObjHeader* newValue, int32_t* spinlock, int32_t* cookie) {
   lock(spinlock);
   ObjHeader* oldValue = *location;
   bool shallRelease = false;
+  bool shallRemember = false;
   // We do not use UpdateRef() here to avoid having ReleaseRef() on return slot under the lock.
   if (oldValue == expectedValue) {
     SetHeapRef(location, newValue);
     shallRelease = oldValue != nullptr;
   } else {
-    if (IsStrictMemoryModel && oldValue != nullptr)
-      rememberNewContainer(oldValue->container());
+    if (IsStrictMemoryModel && oldValue != nullptr) {
+      auto realCookie = computeCookie(oldValue);
+      shallRemember = *cookie != realCookie;
+      if (shallRemember) *cookie = realCookie;
+    }
   }
+  UpdateReturnRef(OBJ_RESULT, oldValue);
   unlock(spinlock);
 
-  UpdateReturnRef(OBJ_RESULT, oldValue);
   if (shallRelease) {
     // No need to rememberNewContainer() on this path, as if `oldValue` is not null - it is explicitly released
     // anyway, and thus can not escape GC.
     ReleaseHeapRef(oldValue);
+  } else {
+    if (IsStrictMemoryModel && oldValue != nullptr && shallRemember) {
+      rememberNewContainer(oldValue->container());
+    }
   }
   return oldValue;
 }
 
-void setHeapRefLocked(ObjHeader** location, ObjHeader* newValue, int32_t* spinlock) {
+
+void setHeapRefLocked(ObjHeader** location, ObjHeader* newValue, int32_t* spinlock, int32_t* cookie) {
   lock(spinlock);
   ObjHeader* oldValue = *location;
   // We do not use UpdateRef() here to avoid having ReleaseRef() on old value under the lock.
   SetHeapRef(location, newValue);
+  *cookie = computeCookie(newValue);
   unlock(spinlock);
   if (oldValue != nullptr)
     ReleaseHeapRef(oldValue);
 }
 
-OBJ_GETTER(readHeapRefLocked, ObjHeader** location, int32_t* spinlock) {
+OBJ_GETTER(readHeapRefLocked, ObjHeader** location, int32_t* spinlock, int32_t* cookie) {
   MEMORY_LOG("ReadHeapRefLocked: %p\n", location)
   lock(spinlock);
   ObjHeader* value = *location;
-  auto* container = value ? value->container() : nullptr;
-  if (container != nullptr)
-    incrementRC<true>(container);
-  unlock(spinlock);
+  auto realCookie = computeCookie(value);
+  bool shallRemember = *cookie != realCookie;
+  if (shallRemember) *cookie = realCookie;
   UpdateReturnRef(OBJ_RESULT, value);
-  if (value != nullptr)
-    ReleaseHeapRef(value);
+  unlock(spinlock);
+#if USE_GC
+  if (IsStrictMemoryModel && shallRemember && value != nullptr) {
+    auto* container = value->container();
+    rememberNewContainer(container);
```

Comment:
Isn't it too late to do this after `unlock`? Another thread might have already nullified the reference and removed its value.